design/src/rfcs/rfc0020_service_builder.md

```diff
@@ -0,0 +1,719 @@
+
+```rust
+async fn handler(Extension(state): Extension</* State */>) -> /* Return Type */ {}
+
+let api_routes = Router::new()
+    .nest("/users", user_routes.layer(Extension(/* state */)))
+    .nest("/teams", team_routes);
+```
+
+In `smithy-rs` a customer is only able to apply a layer to either the `aws_smithy_http::routing::Router` or every route via the [layer method](#router) described above.
+
+# Proposal
+
+The proposal is presented as a series of compatible transforms to the existing service builder, each paired with a motivation. Most of these can be independently implemented, but in the case where there exists an interdependency it is stated.
+
+Although presented as a mutation to the existing service builder, the actual implementation should exist as an entirely separate builder - reusing code generation from the old builder while exposing a new Rust API. Preserving the old API surface will prevent breakage and make it easier to perform comparative benchmarks and testing.
+
+## Remove two-step build procedure
+
+As described in [Builder](#builder), the customer is required to perform two conversions. One from `OperationRegistryBuilder` via `OperationRegistryBuilder::build`, the second from `OperationRegistryBuilder` to `Router` via the `From<OperationRegistry> for Router` implementation. The intermediary stop at `OperationRegistry` is not required and can be removed.
+
+## Statically check for missing Handlers
+
+As described in [Builder](#builder), the `OperationRegistryBuilder::build` method is fallible - it yields a runtime error when one of the handlers has not been set.
+
+```rust
+    pub fn build(
+        self,
+    ) -> Result<OperationRegistry<Op0, In0, Op1, In1>, OperationRegistryBuilderError> {
+        Ok(OperationRegistry {
+            operation0: self.operation0.ok_or(/* OperationRegistryBuilderError */)?,
+            operation1: self.operation1.ok_or(/* OperationRegistryBuilderError */)?,
+        })
+    }
+```
+
+We can do away with fallibility if we allow for on `Op0`, `Op1`, to switch types during build. For example,
+
+```rust
+impl OperationRegistryBuilder<Op0, In0, Op1, In1> {
+    pub fn operation0<NewOp0>(mut self, value: NewOp0) -> OperationRegistryBuilder<NewOp0, In0, Op1, In1> {
+        OperationRegistryBuilder {
+            operation0: value,
+            operation1: self.operation1
+        }
+    }
+    pub fn operation1<NewOp1>(mut self, value: NewOp1) -> OperationRegistryBuilder<Op0, In0, NewOp1, In1> {
+        OperationRegistryBuilder {
+            operation0: self.operation0,
+            operation1: value
+        }
+    }
+}
+
+impl OperationRegistryBuilder<Op0, In0, Op1, In1>
+where
+    Op0: Handler<B, In0, Operation0Input>,
+    Op1: Handler<B, In1, Operation1Input>,
+{
+    pub fn build(self) -> OperationRegistry<Op0, In0, Op1, In1> {
+        OperationRegistry {
+            operation0: self.operation0,
+            operation1: self.operation1,
```

Comment:
I am a bit confused. Doesn't we get a type mismatch with `self.operation0` (an `Option`) and `OperationRegistry.operation0`?

Let me guess. Are you going to propose a solution in type-level state machine?

Reply:
Yep, that's exactly what the intention is here. I should mention that we get rid of `Option`s in addition to simply changing the signature of the setter methods.